draft-ietf-tls-esni.md

```diff
@@ -801,48 +801,37 @@ If the server rejects ECH, the client proceeds with the handshake,
 authenticating for ECHConfig.contents.public_name as described in
 {{auth-public-name}}. If authentication or the handshake fails, the client MUST
 return a failure to the calling application. It MUST NOT use the retry
-configurations.
-
-Otherwise, if both authentication and the handshake complete successfully, the
-client MUST abort the connection with an "ech_required" alert. It then
-processes the "retry_configs" field from the server's "encrypted_client_hello"
-extension.
-
-If at least one of the values contains a version supported by the client, it can
-regard the ECH keys as securely replaced by the server. It SHOULD retry the
-handshake with a new transport connection, using the retry configurations
-supplied by the server. The retry configurations may only be applied to the
-retry connection. The client MUST NOT use retry configurations for connections
-beyond the retry. This avoids introducing pinning concerns or a tracking
-vector, should a malicious server present client-specific retry configurations
-in order to identify the client in a subsequent ECH handshake.
-
-If none of the values provided in "retry_configs" contains a supported version,
-the client can regard ECH as securely disabled by the server. As below, it
-SHOULD then retry the handshake with a new transport connection and ECH
-disabled.
-
-If the field contains any other value, the client MUST abort the connection with
-an "illegal_parameter" alert.
-
-If the server negotiates an earlier version of TLS, or if it does not provide an
-"encrypted_client_hello" extension in EncryptedExtensions, the client proceeds
-with the handshake, authenticating for ECHConfig.contents.public_name as
-described in {{auth-public-name}}. If an earlier version was negotiated, the
-client MUST NOT enable the False Start optimization {{RFC7918}} for this
-handshake. If authentication or the handshake fails, the client MUST return a
-failure to the calling application. It MUST NOT treat this as a secure signal to
+configurations. It MUST NOT treat this as a secure signal to
 disable ECH.
 
-Otherwise, when the handshake completes successfully with the public name
-authenticated, the client MUST abort the connection with an "ech_required"
-alert. The client can then regard ECH as securely disabled by the server. It
-SHOULD retry the handshake with a new transport connection and ECH disabled.
+If the server supplied an "encrypted_client_hello" extension in its
+EncryptedExtensions message, the client MUST check that it is syntactically
+valid and the client MUST abort the connection with a "decode_error" alert
+otherwise. If an earlier TLS version was negotiated, the client MUST NOT enable
+the False Start optimization {{RFC7918}} for this handshake. If both
```

Comment:
This bit seems somewhat out of place -- why include it here? Or include it at all, given the "before sending any application data to the server" clause below?

Reply:
They are not redundant statements.  Forbidding RFC7918 prevents the client from sending application data to the server before the handshake completes.  The additional statement that you suggested ("before sending application data to the server") requires that the connection be aborted before sending any application data to the server after the handshake completes successfully.

I put the RFC7918 statement where it is as it affects what is done (or not) as the handshake progresses, so it is more concise to discuss it prior to everything that follows "If both the authentication and the handshake complete successfully...".  The requirement to validate the syntax of the encrypted_client_hello extension is co-located there under similar reasoning, which is that the syntax must be validated in order to proceed with the "after the handshake" processing that follows (although I think there is currently leeway there for an implementation to postpone the decode until after the handshake completes).

Reply:
> The additional statement that you suggested ("before sending application data to the server") requires that the connection be aborted before sending any application data to the server after the handshake completes successfully.

I don't see how this is the case. The statement says "do not send any application data to the server," which includes, IMO, False Start data, or data that would otherwise be sent when the handshake normally completes.

Reply:
It's the predicate that begins that sentence "If both authentication and the handshake complete successfully,..." that I think places the effect of prohibiting sending application data to the server between the completion of the handshake and the sending of the ech_required alert.

Reply:
It seems redundant, but I can live with it.